File: clients/native/ios/Sources/WuuCore/ThreadSettings.swift
import Foundation

public struct ThreadSettings: Sendable, Equatable {
    public var provider: String
    public var model: String
    public var variant: String
    public var permission: String
    public init(_ thread: JSONValue) {
        provider = thread["model_provider"].string ?? ""
        model = thread["model"].string ?? ""
        variant = thread["model_variant"].string ?? ""
        if variant.isEmpty { variant = thread["model_effort"].string ?? "" }
        permission = thread["permission_mode"].string ?? "standard"
        if permission.isEmpty { permission = "standard" }
    }
    /// A missing thread ID would change the computer's defaults instead of this conversation.
    public func updateParams(threadID: String) throws -> JSONValue {
        guard !threadID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              !provider.isEmpty, !model.isEmpty,
              ["standard", "read_only", "unconfined"].contains(permission) else {
            throw NativeError.invalid("请选择当前会话的模型和权限")
        }
        return ["thread_id": .string(threadID), "provider": .string(provider), "model": .string(model),
                "variant": .string(variant), "permission_mode": .string(permission)]
    }
}

public struct RemoteModel: Identifiable, Sendable {
    public let id: String
    public let name: String
    public let variants: [String]
    init(_ value: JSONValue) {
        id = value["id"].string ?? ""
        name = value["display_name"].string.flatMap { $0.isEmpty ? nil : $0 } ?? id
        let declared = value["variants"].array.compactMap { $0["id"].string }.filter { !$0.isEmpty }
        variants = declared.isEmpty ? value["supported_efforts"].array.compactMap(\.string) : declared
    }
}

public struct RemoteProvider: Identifiable, Sendable {
    public let id: String
    public let models: [RemoteModel]
    public init(_ value: JSONValue) {
        id = value["name"].string ?? ""
        var models = value["models"].array.map(RemoteModel.init).filter { !$0.id.isEmpty }
        if models.isEmpty, let current = value["model"].string, !current.isEmpty {
            models = [RemoteModel(["id": .string(current)])]
        }
        self.models = models
    }
}

File: clients/native/android/app/src/main/java/ai/wuu/nativeapp/ThreadSettings.kt
package ai.wuu.nativeapp

import org.json.JSONObject

data class ThreadSettings(val provider: String, val model: String, val variant: String, val permission: String) {
    // Empty thread_id would mutate computer defaults rather than this conversation.
    fun updateParams(threadID: String): JSONObject {
        require(threadID.isNotBlank() && provider.isNotBlank() && model.isNotBlank() &&
            permission in listOf("standard", "read_only", "unconfined")) { "请选择当前会话的模型和权限" }
        return json("thread_id" to threadID, "provider" to provider, "model" to model,
            "variant" to variant, "permission_mode" to permission)
    }
    companion object {
        fun from(thread: JSONObject) = ThreadSettings(thread.optString("model_provider"), thread.optString("model"),
            thread.optString("model_variant").ifEmpty { thread.optString("model_effort") },
            thread.optString("permission_mode").ifEmpty { "standard" })
    }
}
data class RemoteModel(val id: String, val name: String, val variants: List<String>) {
    companion object {
        fun from(value: JSONObject): RemoteModel {
            val id = value.optString("id")
            val declared = value.optJSONArray("variants")?.objects()?.map { it.optString("id") }?.filter { it.isNotEmpty() }.orEmpty()
            val efforts = value.optJSONArray("supported_efforts")
            return RemoteModel(id, value.optString("display_name").ifEmpty { id },
                declared.ifEmpty { (0 until (efforts?.length() ?: 0)).map { efforts!!.getString(it) } })
        }
    }
}
data class RemoteProvider(val id: String, val models: List<RemoteModel>) {
    companion object {
        fun from(value: JSONObject): RemoteProvider {
            val models = value.optJSONArray("models")?.objects()?.map(RemoteModel::from)?.filter { it.id.isNotEmpty() }.orEmpty()
            val current = value.optString("model")
            return RemoteProvider(value.optString("name"), models.ifEmpty {
                if (current.isEmpty()) emptyList() else listOf(RemoteModel(current, current, emptyList()))
            })
        }
    }
}
